### PR TITLE
feat: support resolving peers with npm aliases

### PR DESCRIPTION
**Pull request by RexSkz**
Background: #4301

My solution is:

- Use the manifest (package.json) to generate a map of NPM aliases.
- When resolving peers, we should check `ctx.parentPkgs[peerName]` and the packages in `npmAliases[peerName]`. Assume we combine the two sources to an array named `pkgsToCheck`.
  - Check each item in `pkgsToCheck`, not only the original `ctx.parentPkgs[peerName]`.
  - During checking, we use a temp var `badPeerIssues` to record the bad peer issues instead of the original `ctx.peerDependencyIssues.bad[peerName]`.
  - If we found a version that satisfies the semver, we found the peer, no need to check more items.
- The result should be one of the following:
  - When we found the peer, nothing bad happens.
  - Else, if there are items in `badPeerIssues`, we add them to `ctx.peerDependencyIssues.bad[peerName]`. We also need to treat the first item of `pkgsToCheck` as the resolved peer (the current version of PNPM will use the `ctx.parentPkgs[peerName]` as the resolved peer).
  - Else, the `missingPeers` should be raised, because we didn't find a satisfying peer and there is no bad peer either.

### CODE REVIEW

**Comment by RexSkz** [comment]
Unfortunately "multiple Jest versions in a single workspace" still doesn't work... These failed tasks (on Windows) seem not to cause by this PR.

image

I rebased the `main` branch and pushed again, but I think they will remain failed.

**Comment by zkochan** [comment]
Yes, this fails on main.

**Comment by zkochan** [comment]
Let's rebase from main and also add a test here: https://github.com/pnpm/pnpm/blob/main/pkg-manager/core/test/install/peerDependencies.ts

**Comment by RexSkz** [comment]
Got it, I'll rebase `main` first, the test will be added later.

**Comment by zkochan** [comment]
What if a package has two versions of webpack? Both satisfy the range of the peer dependency. I think in that case we probably need to prioritize the one that is not aliased.

**Comment by zkochan** [comment]
I don't have a strong opinion about this but it might be not expected to resolve peer dependencies from different versions. After all, peer dependencies are meant to be singletons.

cc @pnpm/collaborators

**Comment by RexSkz** [comment]
> What if a package has two versions of webpack? Both satisfy the range of the peer dependency. I think in that case we probably need to prioritize the one that is not aliased.

It will choose the first one in `pkgsToCheck` which should be the one that is not aliased in most situations, but if both packages are aliases it still chooses an aliased package.

```json
{
  "webpack-51": "npm:webpack@5.1.0",
  "webpack-52": "npm:webpack@5.2.0"
}
```

**Comment by RexSkz** [comment]
As for other package managers, I tried with my private packages:

## NPM 9

```bash
npm i \
  test@3 \
  test-hoc@3 \                 # has peer dependency test@3
  test-next@npm:test@4 \
  test-hoc-next@npm:test-hoc@4 # has peer dependency test@4
```

NPM 9 will throw `ERESOLVE unable to resolve dependency tree` by default, if I use `--legacy-peer-deps` it will generate a flattened `node_modules`, all 4 packages are at the first-level `node_modules`.

```text
node_modules
├─ test
├─ test-hoc
├─ test-next
└─ test-hoc-next
```

It will cause the aliased package `test-hoc-next` use `test` instead of the aliased `test-next`.

## Yarn

### Yarn 1

```bash
yarn add \
  test@3 \
  test-hoc@3 \
  test-next@npm:test@4 \
  test-hoc-next@npm:test-hoc@4
```

Yarn only throws warnings during installation, and the generated `node_modules` is also flattened just like NPM 9.

### Yarn 2 & 3 Plug'n'Play

The virtual store is flattened, and there is not enough information to handle this type of resolution.

Also when I change the `nodeLinker` to `node-modules`, Yarn 2 & 3 produce the same `node_modules` as Yarn 1.

---

So maybe PNPM is the only choice to handle the peers with NPM aliases (thanks to its tree-structured `node_modules`), we can make our own rules in this situation.

**Comment by RexSkz** [comment]
@zkochan The test for `pkg-manager/core/test/install/peerDependencies.ts` is finished. It covers the original issue and the situation when multiple packages satisfy the semver of peer dependency.

File: pkg-manager/core/test/install/peerDependencies.ts [hunk 1]
```diff
@@ -1324,3 +1324,45 @@ test('deduplicate packages that have peers, when adding new dependency in a work
   expect(depPaths).toContain(`/@pnpm.e2e/abc/1.0.0${createPeersFolderSuffix([{ name: '@pnpm.e2e/peer-a', version: '1.0.0' }, { name: '@pnpm.e2e/peer-b', version: '1.0.0' }, { name: '@pnpm.e2e/peer-c', version: '1.0.0' }])}`)
   expect(depPaths).toContain(`/@pnpm.e2e/abc-parent-with-ab/1.0.0${createPeersFolderSuffix([{ name: '@pnpm.e2e/peer-c', version: '1.0.0' }])}`)
 })
+
+test('find correct peer dependency with npm aliases', async () => {
```
Comment (zkochan) [inline pkg-manager/core/test/install/peerDependencies.ts]: It is just aliases, not "npm aliases"
```suggestion
test('find correct peer dependency with aliases', async () => {
```

File: pkg-manager/resolve-dependencies/src/resolvePeers.ts [hunk 2]
```diff
@@ -78,16 +81,30 @@ export function resolvePeers<T extends PartialResolvedPackage> (
...
+      ...manifest.dependencies,
+      ...manifest.devDependencies,
+    }
+    for (const [pkgName, version] of Object.entries(allDependencies)) {
+      if (version.startsWith('npm:')) {
+        const aliasedPkgName = version.substring(4).split('@')[0]
+        npmAliases[aliasedPkgName] = npmAliases[aliasedPkgName] ?? []
+        npmAliases[aliasedPkgName].push(pkgName)
+      }
+    }
```
Comment (zkochan) [inline pkg-manager/resolve-dependencies/src/resolvePeers.ts]: The way you did it, aliased packages will be used to resolve peer dependencies only if they are direct dependencies of the project. But aliases should resolve peers from any parent dependencies. See how `parentPkgs` is extended with more dependencies every time we walk the graph deeper.

File: pkg-manager/resolve-dependencies/src/resolvePeers.ts [hunk 3]
```diff
@@ -78,16 +81,30 @@ export function resolvePeers<T extends PartialResolvedPackage> (
...
     }
 
+    const npmAliases: Record<string, string[]> = {}
+    const allDependencies = {
+      ...manifest.dependencies,
+      ...manifest.devDependencies,
+    }
+    for (const [pkgName, version] of Object.entries(allDependencies)) {
+      if (version.startsWith('npm:')) {
+        const aliasedPkgName = version.substring(4).split('@')[0]
```
Comment (zkochan) [inline pkg-manager/resolve-dependencies/src/resolvePeers.ts]: I don't think there's a need to parse this info. The package name and alias name are already present in the nodes of dependenciesGraph

**Review by zkochan — CHANGES_REQUESTED** [review]

File: pkg-manager/resolve-dependencies/src/resolvePeers.ts (hunk 3 above)
Comment (RexSkz) [reply pkg-manager/resolve-dependencies/src/resolvePeers.ts]: Although we can get the values in `dependenciesGraph`, there're still two problems:

- `dependenciesGraph` is an object, which means we have to loop through its key to find all dependencies starting with `peerName`, which destroys the performance.
- Even if the performance slowdown is acceptable, there is no alias info in the value in `dependenciesGraph` (the field `name` and `version` look as if it's not aliased).

File: pkg-manager/resolve-dependencies/src/resolvePeers.ts (hunk 2 above)
Comment (RexSkz) [reply pkg-manager/resolve-dependencies/src/resolvePeers.ts]: `ctx.parentPkgs` is a map with "aliased name" -> "node", but in `_resolvePeers` we can only get the unaliased name, so it couldn't be used directly.

```ts
for (const peerName in ctx.resolvedPackage.peerDependencies) { ... }
```

I generate `npmAliases` to let it help me find the items in `ctx.parentPkgs`. It does resolve peers from any parent dependencies:

```ts
const pkgsToCheck = [
  ctx.parentPkgs[peerName],
  ...(ctx.npmAliases[peerName] || []).map((alias) => ctx.parentPkgs[alias]),
].filter(Boolean);
```

File: pnpm/bin/pnpm.cjs [hunk 4]
```diff
@@ -13,7 +13,7 @@ ${COMPATIBILITY_PAGE}`)
 }
 
 global['pnpm__startedAt'] = Date.now()
-require('../dist/pnpm.cjs')
+// require('../dist/pnpm.cjs')
 
 // if you want to debug at your local env, you can use this
-// require('../lib/pnpm')
+require('../lib/pnpm')
```
Comment (kenrick95) [inline pnpm/bin/pnpm.cjs]: Is this change accidental?

Comment (RexSkz) [reply pnpm/bin/pnpm.cjs]: Oh I forgot to revert it...🤦

Comment (RexSkz) [reply pnpm/bin/pnpm.cjs]: It should be reverted now.

**Comment by zkochan** [comment]
I have merged the v8 branch to main, so the `peerDependencies.ts` tests will fail because they use the old lockfile format.

**Comment by RexSkz** [comment]
> I have merged the v8 branch to main, so the `peerDependencies.ts` tests will fail because they use the old lockfile format.

Should I rebase `main` and update the test case?

BTW, is v7 still under maintenance, or the new releases will all be v8?

**Comment by zkochan** [comment]
Yes, rebase

I don't think that this will be released in v7. It is too big change.

v7 will get bug fixes but no new features

**Comment by RexSkz** [comment]
Rebase done~

**Comment by RexSkz** [comment]
The action log shows an exit code `3221225477` (`0xc0000005`), which refers to the `Access Violation Error` in Windows.

image

Could this PR cause the error? I'm unsure and don't know how to fix it. (I did some searching but only to find it happens occasionally and can't be fixed with pure JavaScript.)

**Comment by zkochan** [comment]
Probably unrelated, though I haven't seen such error in the past.

**Comment by zkochan** [comment]
I have pushed a test that shows I mean by aliased subdep not resolving peer dependencies.

**Comment by zkochan** [comment]
Here's another case that bothers me.

```
react@16
react17@npm:react@17
card
  react@16
  button (has react@^17 in peers)
    react@16 should still use react 16 because it is declared by a closer parent package
```

**Comment by RexSkz** [comment]
> Here's another case that bothers me.
> 
> ```
> react@16
> react17@npm:react@17
> card
>   react@16
>   button (has react@^17 in peers)
>     react@16 (should still use react 16 because it is declared by a closer parent package)
> ```

After this PR, it will (wrongly) become `react@17` because it satisfies the peers of `button`, which needs to be fixed.

In my thought, since...

1. a package in `dependencies` should have a higher priority than the package with the same name in `peerDependencies`,
1. peers don't affect the generation of `node_modules` in both NPM and Yarn,

...maybe we should add this situation to the rule in `_resolvePeers`:

```js
// the first loop, newly added logic, found a proper version according to dependencyVersionRange
// and record the peer issues
for (const resolved of pkgsToCheck) {

  // we should also pass `dependencies`, currently there is no "ctx.resolvedPackage.dependencies"
  const dependencyVersionRange = ctx.resolvedPackage.dependencies[peerName]
  const peerVersionRange = ctx.resolvedPackage.peerDependencies[peerName].replace(/^workspace:/, '')

  // it satisfies dependency semver
  if (semverUtils.satisfiesWithPrereleases(resolved.version, dependencyVersionRange, true)) {
    resolvedPeers[peerName] = resolved.nodeId
    foundPeer = true

    // but it not satisfies peer semver, so it's a "bad peer"
    if (semverUtils.satisfiesWithPrereleases(resolved.version, peerDependencyVersionRange, true)) {
      badPeerIssues.push(...)
    }

    break
  }
}

// the second loop, found a proper version according to peerVersionRange
// and record the peer issues
if (!foundPeer) {
  for (const resolved of pkgsToCheck) { ... }
}

// deal with the recorded "bad peers" or "missing peers"
if (!foundPeer) {
  ...
}
```

**Comment by zkochan** [comment]
> > Here's another case that bothers me.
> > ```
> > react@16
> > react17@npm:react@17
> > card
> >   react@16
> >   button (has react@^17 in peers)
> >     react@16 (should still use react 16 because it is declared by a closer parent package)
> > ```
> 
> After this PR, it will (wrongly) become `react@17` because it satisfies the peers of `button`, which needs to be fixed.
> 
> In my thought, since...
> 
> 1. a package in `dependencies` should have a higher priority than the package with the same name in `peerDependencies`,
> 2. peers don't affect the generation of `node_modules` in both NPM and Yarn,

`react@16` is in the dependencies of card (not in the peer dependencies). It should be used to resolve the peer dependency of button.

***

The issue that this PR aims to solve is only about a single dependency that is aliased. I think we should fix just that. Users should never install two different versions of a peer dependency in one project. It doesn't make sense to do so as peer dependencies should be singletons.

But if someone for some reason does want to use two versions of a peer dependency, they may use [`packageExtensions`](https://pnpm.io/package_json#pnpmpackageextensions) to add a different version of a peer dependency as a regular dependency to a package that needs a different version of it.

**Comment by RexSkz** [comment]
Oh I got your idea, will try to fix the new test case using `parentRefs`.

**Comment by zkochan** [comment]
Maybe you should create a new PR in which you remove the `for (const resolved of pkgsToCheck) {` loop. I don't think we will allow to resolve peers from multiple dependencies this way.

**Comment by RexSkz** [comment]
> Maybe you should create a new PR in which you remove the `for (const resolved of pkgsToCheck) {` loop. I don't think we will allow to resolve peers from multiple dependencies this way.

I don't get it... How to choose one that satisfies version range restriction from multiple peers without using loop?

**Comment by zkochan** [comment]
You just pick one dependency to resolve peer dependencies. If there is an aliased one and a non-aliased one, you pick the non-aliased one. If there is only an aliased one, then you pick it. If there are multiple non-aliased ones, then maybe pick the highest one.

So basically, you keep the peer dependency resolution part as is. You only change the parent pkgs object, so that it is a map of aliases to versions and also real package names to versions.

**Comment by RexSkz** [comment]
Alright, I'll pick one and validate its semver as before. It does keep the result stable if other non-relative packages are upgraded.

BTW, I found a test case failed: [external lockfile: peer dependency is grouped with dependent even after a named update of the resolved package](https://github.com/RexSkz/pnpm/blob/rex/resolve-peers-with-npm-aliases/pkg-manager/core/test/install/peerDependencies.ts#L865-L908), the steps are:

1. prepareEmpty()
2. add dependency `@pnpm.e2e/peer-c@1.0.0` and `@pnpm.e2e/abc-parent-with-ab@1.0.0`
3. add dependency `@pnpm.e2e/peer-c@2.0.0`

The reason this PR fails on it: in `_resolvePeers`, we have two `peer-c` (1.0.0 and 2.0.0) in `pkgsToCheck`, but it should be only `2.0.0`. The root cause is that `topParents` is generated from `node_modules`, it will remain `1.0.0` before step 3 is done.

Is there a proper way to let `_resolvePeers` know `1.0.0` should not be picked?

**Comment by RexSkz** [comment]
I found a problem in this case if we just choose one peer instead of trying each `pkgsToCheck`:

```c
ajv-old@npm:ajv@6.0.0
ajv@7.0.0
ajv-keywords@3.0.0 // with peer ajv@^6
```

The peers for `ajv-keywords` will be `ajv@7.0.0` (the higher version), which does not satisfy the peer version range in `ajv-keywords`.

It can not be solved by changing the rule of peer selection.

**Comment by zkochan** [comment]
It is fine. You cannot solve all variations of peer dependency issues automatically.

**Comment by RexSkz** [comment]
Alright, this PR should just fix #4301 and nothing more. If users need to use more than one version of peers, they need to specify `packageExtensions` manually.

For example, the `ajv` issue can be solved after this PR:

```json
{
  // the requirement
  "dependencies": {
    "ajv-old": "npm:ajv@6.0.0",
    "ajv-keywords-old": "npm:ajv-keywords@3.0.0",
    "ajv": "7.0.0",
    "ajv-keywords": "4.0.0"
  },
  // add an aliased dependency for ajv-keywords@3 (ajv-keywords-old)
  "pnpm": {
    "packageExtensions": {
      "ajv-keywords@3": {
        "dependencies": {
          "ajv-old": "npm:ajv@6.0.0"
        }
      }
    }
  }
}
```

the lock file will be:

```yaml
packages:

  /ajv-keywords@3.0.0(ajv@6.0.0):
    resolution: {integrity: sha512-uMRktPTgqPL+jx4aRg29iv9+jNPDH3NGw4tn/dKONgwd53XPdmZe433n4eyjsj92v0Koe0qqHwAodZCzIriwrQ==}
    peerDependencies:
      ajv: ^6.0.0
    dependencies:
      ajv: 6.0.0
      ajv-old: /ajv@6.0.0
    dev: false

  /ajv-keywords@4.0.0(ajv@7.0.0):
    resolution: {integrity: sha512-baL4pEYniCF5E/5Cj28f1DmPXGGASQIeCFfntY94vJPtrq0fei3iNt/TP5f2IwEH4opCzcOOvL6hKsi2IHaecg==}
    peerDependencies:
      ajv: ^7.0.0
    dependencies:
      ajv: 7.0.0
    dev: false
```

File: pkg-manager/resolve-dependencies/src/resolvePeers.ts [hunk 5]
```diff
@@ -182,30 +182,39 @@ function createPkgsByName<T extends PartialResolvedPackage> (
...
+      version,
+      nodeId: linkedDir,
+    }
+    if (version.startsWith('npm:')) {
+      const index = version.lastIndexOf('@')
+      const aliasedPkgName = version.slice(4, index)
+      const aliasedPkgVersion = version.slice(index + 1)
+      pkg.alias = aliasedPkgName
+      updateParentRefs(parentRefs, aliasedPkgName, aliasedPkgVersion, pkg)
+    } else if (alias) {
```
Comment (RexSkz) [inline pkg-manager/resolve-dependencies/src/resolvePeers.ts]: Oh, this condition is redundant; I'll remove it after CI finishes :)

File: pkg-manager/core/test/install/peerDependencies.ts [hunk 6]
```diff
@@ -1412,7 +1412,7 @@ test('resolve peer dependency from aliased direct dependency', async () => {
   await addDependenciesToPackage(manifest, ['@pnpm.e2e/abc@1.0.0'], opts)
 
   const lockfile = await readYamlFile<any>(path.resolve(WANTED_LOCKFILE)) // eslint-disable-line
-  expect(lockfile.packages).toHaveProperty(['/@pnpm.e2e/abc@1.0.0(@pnpm.e2e/peer-a@1.0.0)'])
+  expect(lockfile.packages['/@pnpm.e2e/abc@1.0.0(@pnpm.e2e/peer-a@1.0.0)']).toBeTruthy()
```
Comment (zkochan) [inline pkg-manager/core/test/install/peerDependencies.ts]: why was this changed?

Comment (RexSkz) [reply pkg-manager/core/test/install/peerDependencies.ts]: We shouldn't use `toHaveProperty` if there's a dot in the property, it'll be treated as nested property such as `user.name`.

Comment (zkochan) [reply pkg-manager/core/test/install/peerDependencies.ts]: I don't think so because it works

Comment (zkochan) [reply pkg-manager/core/test/install/peerDependencies.ts]: For `user.name` you'd do `.toHaveProperty(['user', 'name'])`

Comment (RexSkz) [reply pkg-manager/core/test/install/peerDependencies.ts]: It doesn't work when I run tests on my PC:

```text
  ● when there are several aliased dependencies of the same package, pick the one with the highest version to resolve peers                                                                                                                                                                                                                                      
                                                                                                                                                                                                                                                                                                                                                                 
    expect(received).toHaveProperty(path)

    Expected path: "/@pnpm.e2e/abc@1.0.0(@pnpm.e2e/peer-c@2.0.0)"
    Received path: []

    Received value: {"/@pnpm.e2e/abc@1.0.0(@pnpm.e2e/peer-c@2.0.0)": {"dependencies": {"@pnpm.e2e/dep-of-pkg-with-1-dep": "100.0.0", "@pnpm.e2e/peer-c": "2.0.0"}, "dev": false, "peerDependencies": {"@pnpm.e2e/peer-a": "^1.0.0", "@pnpm.e2e/peer-b": "^1.0.0", "@pnpm.e2e/peer-c": "^1.0.0"}, "resolution": {"integrity": "sha512-/BCMmCRQ58DLWtF5iU97Lbf0s+llpWHvvZvQfFnCYFBpa1pfYnlJW1k6+Z4UzcUdYrUvyDmYPwxGVvPStAUw1w=="}}, "/@pnpm.e2e/dep-of-pkg-with-1-dep@100.0.0": {"dev": false, "resolution": {"integrity": "sha512-KUmwlEiE3pzpmPCfrLkoOO7fmL4Tft+dcm6YQlUfOUJvHO73xlcKVoA/xOHP/ayUO8GiPlIqWRbKHrsxvXGE8g=="}}, "/@pnpm.e2e/peer-c@1.0.0": {"dev": false, "resolution": {"integrity": "sha512-SQ3TfmnPopiVIqzC/PVV8WlfJjVWFdszQhfZG6e9stDFD9i9dZ5VQImKYhvJsME4nb45rHGsc/4CmroEjO7aCA=="}}, "/@pnpm.e2e/peer-c@1.0.1": {"dev": false, "resolution": {"integrity": "sha512-6ag+AQOLJgsioiG8aInnOJGautKz7lNf3zKQ8FJHOb4b3yIclmUofTaDwth9ydp9W56qzwbHWXjkdrqfW5JViQ=="}}, "/@pnpm.e2e/peer-c@2.0.0": {"dev": false, "resolution": {"integrity": "sha512-s0TPtcya7A2r7vSzmq1dNAdxUNY0TQXeyARH8YMe/d7wb3/I4gryxBP3VVGPuKf6LUO+AALCBr0Jha8wwG7isg=="}}}

      1424 |
      1425 |   const lockfile = await readYamlFile<any>(path.resolve(WANTED_LOCKFILE)) // eslint-disable-line
    > 1426 |   expect(lockfile.packages).toHaveProperty('/@pnpm.e2e/abc@1.0.0(@pnpm.e2e/peer-c@2.0.0)')
           |                             ^
      1427 | })
      1428 |

      at Object.<anonymous> (test/install/peerDependencies.ts:1426:29)
```

The [Jest docs](https://jestjs.io/docs/expect#tohavepropertykeypath-value) says, "For checking deeply nested properties in an object you may use [dot notation](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Operators/Property_accessors) or an array containing the keyPath for deep references".

**Review by zkochan — APPROVED** [review]
